Script first commit

MarineGirardey opened the pull request:
Script generating DataFrames from Chembl id of the molecules from Drug OT parquet files to PDB id, structure ID linked to these molecule and their associated target gene id.

MarineGirardey committed:
Script first commit

MarineGirardey committed:
Add script.py

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+
+parser = argparse.ArgumentParser(add_help=True, description=program_description)
+
+parser.add_argument('-p', '--parquet', help='Path to parquet folder', default=None, metavar='parquet_file_path', type=str, required=True)
+parser.add_argument('-u', '--unichem', help='Path to unichem src1src3.txt file', default=None, metavar='unichem_file_path', type=str, required=True)
+parser.add_argument('-e', '--ensembl', help='Path to pdb_chain_ensembl.csv file', default=None, metavar='ensembl_file_path', type=str, required=True)
```
Just for the record it would be great to add as comments where these datafiles are available from.

MarineGirardey replied:
```python

program_description = '''
This script retrieves the structures in complex with the molecules from parquet.
How to find the files?
molecule file here (Drug): http://ftp.ebi.ac.uk/pub/databases/opentargets/platform/21.11/output/etl/parquet/molecule/*
unichem file : https://ftp.ebi.ac.uk/pub/databases/chembl/UniChem/data/wholeSourceMapping/src_id1/src1src3.txt.gz
ensembl file : https://ftp.ebi.ac.uk/pub/databases/msd/sifts/csv/pdb_chain_ensembl.csv
'''

parser = argparse.ArgumentParser(add_help=True, description=program_description)

parser.add_argument('-p', '--parquet', help='Path to parquet folder', default=None, metavar='parquet_file_path', type=str, required=True)
parser.add_argument('-u', '--unichem', help='Path to unichem src1src3.txt file', default=None, metavar='unichem_file_path', type=str, required=True)
parser.add_argument('-e', '--ensembl', help='Path to pdb_chain_ensembl.csv file', default=None, metavar='ensembl_file_path', type=str, required=True)

args = parser.parse_args()

```

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+def create_pdb_target_gene_df(path_id_file, unichem_molecule_struct_spark_df):
+
+    pdb_chain_ensembl = spark.read.csv(path_id_file, sep=',', header=True, comment='#')
+    pdb_chain_ensembl = pdb_chain_ensembl.select('PDB', 'GENE_ID').distinct()
```
This two step operation can be made easier to understand using the following expression:

```suggestion
pdb_chain_ensembl = (
    spark.read.csv(path_id_file, sep=',', header=True, comment='#')
    .select('PDB', 'GENE_ID')
    .distinct()
)
```

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+                                       .join(exploded_df, pdb_chain_ensembl["PDB"] == exploded_df["col"])
+                                       .drop('col'))
+
+    return chain_ensembl_struct_mol_joined
```
It might be a bit much to take, but these expressions should be pooled together like this:
```suggestion
return (
    unichem_molecule_struct_spark_df
    
    # Exploding the STRUCTURE_ID array into a column called PDB:
    .select(
        'MOLECULE_PDB_ID','MOLECULE_CHEMBL_ID',
        f.explode(col('STRUCTURE_ID')).alias('PDB')
    )
    
    # Joining chain ids by PDB identifier:
    .join(pdb_chain_ensembl, on='PDB', how='inner')
    
    # The dataframe is stored in memory when returning:
    .persist()
)
```

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+        spark.read
+        .parquet(args.parquet)
+        .select(f.col('id'))
+        .withColumnRenamed('id', 'MOLECULE_CHEMBL_ID')
```
You can rename it with one step, but I would also select the molecule name, as knowing the name, not just the ID would help a lot:
```suggestion
        .select(f.col('name'), f.col('id').alias('MOLECULE_CHEMBL_ID'))
```

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+
+    # ----- GET PDB ID -----
+    # Unichem molecules DataFrame
+    unichem_df = spark.read.csv(args.unichem, sep=r'\t', header=True)
```
I would do the column renaming here:
```suggestion
unichem_df = (
    spark.read.csv(unichem, sep=r'\t', header=True)
    .withColumnRenamed('From src:\'1\'', 'CHEMBL_MOLECULE_ID')
    .withColumnRenamed('To src:\'3\'', 'PDB_COMPOUND_ID')
)
```

Also, when renaming the IDs, it makes sense this time adding the source and the type of the identifier as well. eg we have compound identifiers from PDB and we have structure identifiers from pdb.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+
+    # Join Unichem and Molecule DataFrame
+    unichem_molecule_df = (molecule_df
+                           .join(unichem_df, unichem_df["From src:'1'"] == molecule_df['MOLECULE_CHEMBL_ID'])
```
You don't need to drop the column you joined the two tables with if you are already renamed them. And in general it is easier to deal with the columns if you rename the columns when you read the files.

MarineGirardey replied:
#### This is what I have if I dont select/drop a column:

```
+-----------------+------------------+------------------+---------------+
|             name|CHEMBL_MOLECULE_ID|CHEMBL_MOLECULE_ID|PDB_COMPOUND_ID|
+-----------------+------------------+------------------+---------------+
|       MIGALASTAT|      CHEMBL110458|      CHEMBL110458|            DGJ|
|   (S)-Fluoxetine|     CHEMBL1169388|     CHEMBL1169388|            SFX|
|        RUCAPARIB|     CHEMBL1173055|     CHEMBL1173055|            RPB|
|    CHEMBL1231606|     CHEMBL1231606|     CHEMBL1231606|            C4F|
|    CHEMBL1232381|     CHEMBL1232381|     CHEMBL1232381|            DTB|
|      PF-04691502|     CHEMBL1234354|     CHEMBL1234354|            ML9|
```

#### I replaced the drop by:

```python
# Join Unichem and Molecule DataFrame
unichem_molecule_df = (molecule_df.alias("a")
                       .join(unichem_df.alias("b"), unichem_df["CHEMBL_MOLECULE_ID"] == molecule_df['CHEMBL_MOLECULE_ID'])
                       ).select("name", "a.CHEMBL_MOLECULE_ID", "PDB_COMPOUND_ID")
```

Is it better?

DSuveges replied:
If you join two tables by columns with the same name you can simplify this expression:
```python
unichem_df["CHEMBL_MOLECULE_ID"] == molecule_df['CHEMBL_MOLECULE_ID']
```
to this:
```python
on='CHEMBL_MOLECULE_ID'
```
In the above case columns will not be duplicated. Also if you need to join by multiple columns you can say:
```python
on=['CHEMBL_MOLECULE_ID', 'targetId']
```

Nicer, isn't it?

MarineGirardey replied:
Much better!! Thank you

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+                                        .orderBy(f.desc("NB_OF_STRUCT_PER_MOL"))
+                                        )
+
+    unichem_molecule_struct_spark_df = (unichem_molecule_struct_spark_df
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL692')
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL1236970')
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL457299')
+                                        .filter(f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL113178')
+                                        .select('MOLECULE_CHEMBL_ID', 'MOLECULE_PDB_ID', 'STRUCTURE_ID'))
```
Multiple filters should be collapsed into a single expression:

```suggestion
unichem_molecule_struct_spark_df = (
    unichem_molecule_struct_spark_df
    .filter(
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL692') & 
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL1236970') &
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL457299') & 
        (f.col('MOLECULE_CHEMBL_ID') != 'CHEMBL113178')
    )
    .select('MOLECULE_CHEMBL_ID', 'MOLECULE_PDB_ID', 'STRUCTURE_ID')
)
```

However we should create a filter array. You can imagine how complicated this experssion would be if there would 20 molecules to exclude.

```python
dropped_molecules = [
    'CHEMBL692', 'CHEMBL1236970', 'CHEMBL457299', 'CHEMBL113178'
]

unichem_molecule_struct_spark_df = (
    unichem_molecule_struct_spark_df
    .filter(~col('MOLECULE_CHEMBL_ID').isIn(dropped_molecules))
    .select('MOLECULE_CHEMBL_ID', 'MOLECULE_PDB_ID', 'STRUCTURE_ID')
)
```

MarineGirardey replied:
`isin` worked for me instead of `isIn`

DSuveges reviewed (changes requested):
I couldn't finish with the review, but please take a look and try to address them.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+        pdb_id: id of the molecule for whom we want to know the structures associated
+    --------------
+    Return
+        data[pdb_id] : data is a dictionary, we want the value which is the list of structures
```
I like the detailed docstring of this function. However writing a good docstring is always easier for a function which is easy to understand already. In this case the function is just a few lines. And its function is quite intuitive. Also the return value of this function is a list not a dictionary, or at least this is how it seems form the text. When writing documentation, always try to be short but very precise.

How I would re-phrase this docstring:
```python
def get_structure(pdb_compound_id: str) -> list:
    """Fetching structure identifiers from PDBkb REST API

    Args:
        pdb_compound_id: string, a single compound identifier
    Returns:
        List of PDB structure identifiers where the compound can be found
    """
```

What you see here:
* The input parameter's name is more telling. It's not just any identifier from PDB. This is a compound id.
* You can specify the type of the input parameters by using the `:` and the return type by `->`. This serves only documentation purposes. It's not enforced by anyhow.
* The docstring follows [google guidelines](https://google.github.io/styleguide/pyguide.html). In general, reading the style guide is a good time investment.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+            return []
+
+
+def create_pdb_target_gene_df(path_id_file, unichem_molecule_struct_spark_df):
```
Writing a good docstring here would be very helpful. This function is not intuitive.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+    # ----- GET STRUCTURE ID -----
+    # Apply get_structure function on the Unichem-Molecule DataFrame
+
+    unichem_molecule_struct_pd_df = (unichem_molecule_df
```
Yes, using a function where you requests data from an API, it makes sense to convert it to pandas dataframe. I like to see the `.assign` method being used. It can be further optimized a bit, but it's good.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+
+
+if __name__ == '__main__':
+    main = main()
```
Oh, this is really bad. Now you are overwriting the name `main` (the function) with `main` the variable. In general I would recommend not use the name `main` for variable given how consistently it is used to denote the main function.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+    nb_target_pd = pdb_target_df.count()
+    nb_human_target = pdb_target_df.filter(pdb_target_df.GENE_ID.startswith('ENSG')).count()
+
+    count_distinct_target_pd_df = pdb_target_df.groupBy('GENE_ID').count().toPandas()
```
This variable is not used.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,186 @@
+        'STRUCTURE_ID').distinct().count()
+
+    nb_target_pd = pdb_target_df.count()
+    nb_human_target = pdb_target_df.filter(pdb_target_df.GENE_ID.startswith('ENSG')).count()
```
When referring to columns it is nicer to use the `col` function eg:
```suggestion
    nb_human_target = pdb_target_df.filter(col('GENE_ID').startswith('ENSG')).count()
```

MarineGirardey replied:
#### Ok thank you! Do you know why it's better?

DSuveges replied:
It's better, not necessarily for better performance, but your expression will become you won't need to repeat the name of the dataframe. (Think about changing the name of the dataframe)

MarineGirardey committed:
Daniel review applied on script.py

Improvment changes, do not change the output

DSuveges commented on script.py:
```diff
@@ -0,0 +1,320 @@
+molecule_path = '/Users/marinegirardey/Documents/OpenTargetInternship/molecule/'
+# molecule = '/Users/marinegirardey/Documents/OpenTargetInternship/small_folder/'
+unichem_path = '/Users/marinegirardey/Documents/OpenTargetInternship/id_files/src1src3.txt'
+ensembl_path = '/Users/marinegirardey/Documents/OpenTargetInternship/chain_ensembl_struct_mol_joined.csv'
```
If you are parsing arguments (at line 33-37), why are these values are hardcoded?

DSuveges commented on script.py:
```diff
@@ -0,0 +1,320 @@
+
+
+# TIMER 1 START
+start_1 = timeit.default_timer()
```
You don't need to use timer. Instead use a properly initialized logger. Please take a look at how we manage logging [here](https://github.com/opentargets/evidence_datasource_parsers/blob/master/modules/TEP.py) for example. Then, when you write to the logs, timestamps will be added automatically.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,320 @@
+        .filter(~f.col('CHEMBL_MOLECULE_ID').isin(dropped_molecules))
+        .select('*')
+)
+unichem_molecule_struct_spark_df.show()
```
In the production version of the script please don't use `.show()`. At this point we should know what is in these dataframes. You can build in assertions, and if the assertions violated, you can show the header, but that should be part of the troubleshooting.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,320 @@
+# In[ ]:
+
+
+print('----STATISTICS----')
```
These stats should be written out as logs. Eg. when you have read the molecules dataset, you can log:
```python
logging.info(f"Number of molecules in the OpenTargets drug index: {molecules.count()}")
```

DSuveges reviewed (changes requested):
Getting better. There are a few recommendations + two general tip:

* Please use the `if __name__ == '__main__':` pattern. As a rule of thumb, it makes sense to keep every line of code of the script in function.
* Please remove the excessive blank lines. One blank lines are fine. Also the comment lines that doesn't really say anything (`# In[ ]:`)

MarineGirardey committed:
script.py output updated and improvement

- Used Inchikey instead of Unichem
- Taking only one structure per molecule-target complex
- Daniel advices ameliorations

DSuveges commented on script.py:
```diff
@@ -0,0 +1,269 @@
+
+    logging.info('FINISHED: Get pdbStructureId on pdbe API.')
+
+    inchikey_molecule_structure_joined_pd_df.to_csv('output_files/structure_of_molecules.csv', index=False, header=True)
```
Any output generated should either be hardcoded or provided by a command line argument should not expect the presence of a folder. As I didn't have the `output_files` on my computer, this script failed at this point. 

```
Traceback (most recent call last):
  File "/Users/dsuveges/repositories/drugDataEnrichment/script.py", line 285, in <module>
    main()
  File "/Users/dsuveges/repositories/drugDataEnrichment/script.py", line 84, in main
    inchikey_molecule_structure_joined_pd_df.to_csv('output_files/structure_of_molecules.csv', index=False, header=True)
  File "/Users/dsuveges/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pandas/core/generic.py", line 3387, in to_csv
    return DataFrameRenderer(formatter).to_csv(
  File "/Users/dsuveges/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pandas/io/formats/format.py", line 1083, in to_csv
    csv_formatter.save()
  File "/Users/dsuveges/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pandas/io/formats/csvs.py", line 228, in save
    with get_handle(
  File "/Users/dsuveges/opt/anaconda3/envs/jupyter3.9/lib/python3.9/site-packages/pandas/io/common.py", line 647, in get_handle
    handle = open(
FileNotFoundError: [Errno 2] No such file or directory: 'output_files/structure_of_molecules.csv'
```

However I would only output the final dataset no intermedier ones. Also it is highly recommended to not hardcode the output filename but provide as command line argument.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,269 @@
+    inchikey_molecule_structure_joined_pd_df.to_csv('output_files/structure_of_molecules.csv', index=False, header=True)
+
+    # Convert Spark DataFrame into Pandas DataFrame
+    inchikey_molecule_structure_joined_spark_df = spark.createDataFrame(inchikey_molecule_structure_joined_pd_df)
```
I would simplify the above expression because the pandas dataframe is not used at all.

```python
    inchikey_molecule_joined = (
        spark.createDataFrame(
            (
                inchikey_molecule_joined_df
                .toPandas()
                .assign(pdbStructureId=lambda df: df.pdbCompoundId.apply(get_structure))
            )
        )
    )
```

DSuveges commented on script.py:
```diff
@@ -0,0 +1,269 @@
+
+    logging.info(f'RESULT: Number of molecules in Inchikey (inchikeyId): {inchikey_df.count()}')
+
+    # Join Unichem and Molecule DataFrame
```
Always be very considerate about the comments! The comments are not evolving automatically with the code, they can be out of date, which will confuse the next person touching the code.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,269 @@
+        .join(pdb_target_df, on="pdbStructureId", how="inner") # Only keep the structures mapped by ensembl (inner)
+    )
+
+    df = (
```
Please always try to avoid such variable names. I see certain cases it is not easy to find the proper variable name to reflect the content/function, but if all we know about the structure that it is a dataframe... well, that's hard to follow.

DSuveges commented on script.py:
```diff
@@ -0,0 +1,269 @@
+        return []
+
+
+def create_pdb_target_gene_df(path_id_file: str, unichem_molecule_struct_df):
```
It seems, this function is not used. What was the plan with this guy?

DSuveges reviewed (changes requested):
Things are definitely getting better. Just make sure all functions and variables are used, for checking this and other formatting related issues, please use a python linter on your IDE eg [Flake8](https://youtrack.jetbrains.com/issue/PY-41427). Other linters could work too. Please check pycharm documentations on how to enable these.

MarineGirardey committed:
add output_file folder

Commit:
plip script added

Commit:
script to compute genomic position of each residue interacting in each structures
